agent/logger/log.go

```diff
@@ -14,87 +14,162 @@
+	logfile       string
+	level         string
+	outputFormat  string
+	lock          sync.Mutex
+}
 
-// Initialize this logger once
-var once sync.Once
+var Config *logConfig
 
-func init() {
-	once.Do(initLogger)
+func logfmtFormatter(params string) seelog.FormatterFunc {
+	return func(message string, level seelog.LogLevel, context seelog.LogContextInterface) interface{} {
+		return fmt.Sprintf(`level=%s time=%s msg=%q module=%s
+`, level.String(), time.Now().UTC().Format(time.RFC3339), message, context.FileName())
+	}
 }
 
-func initLogger() {
-	levels = map[string]string{
-		"debug": "debug",
-		"info":  "info",
-		"warn":  "warn",
-		"error": "error",
-		"crit":  "critical",
-		"none":  "off",
+func jsonFormatter(params string) seelog.FormatterFunc {
+	return func(message string, level seelog.LogLevel, context seelog.LogContextInterface) interface{} {
+		return fmt.Sprintf(`{"level": %q, "time": %q, "msg": %q, "module": %q}
+`, level.String(), time.Now().UTC().Format(time.RFC3339), message, context.FileName())
 	}
-
-	level = DEFAULT_LOGLEVEL
-
-	logger = &Shim{}
-
-	envLevel := os.Getenv(LOGLEVEL_ENV_VAR)
-
-	logfile = os.Getenv(LOGFILE_ENV_VAR)
-	SetLevel(envLevel)
-	registerPlatformLogger()
-	reloadConfig()
 }
 
 func reloadConfig() {
-	logger, err := log.LoggerFromConfigAsString(loggerConfig())
+	logger, err := seelog.LoggerFromConfigAsString(seelogConfig())
 	if err == nil {
-		log.ReplaceLogger(logger)
+		seelog.ReplaceLogger(logger)
 	} else {
-		log.Error(err)
+		seelog.Error(err)
 	}
 }
 
+func seelogConfig() string {
+	c := `
+<seelog type="asyncloop" minlevel="` + Config.level + `">
+	<outputs formatid="` + Config.outputFormat + `">
+		<console />`
+	c += platformLogConfig()
+	if Config.logfile != "" {
+		if Config.RolloverType == "size" {
+			c += `
+		<rollingfile filename="` + Config.logfile + `" type="size"
+		 maxsize="` + strconv.Itoa(int(Config.MaxFileSizeMB*1000000)) + `" archivetype="none" maxrolls="` + strconv.Itoa(Config.MaxRollCount) + `" />`
+		} else {
+			c += `
+		<rollingfile filename="` + Config.logfile + `" type="date"
+		 datepattern="2006-01-02-15" archivetype="none" maxrolls="` + strconv.Itoa(Config.MaxRollCount) + `" />`
+		}
+	}
+	c += `
+	</outputs>
+	<formats>
+		<format id="logfmt" format="%EcsAgentLogfmt" />
+		<format id="json" format="%EcsAgentJson" />
+	</formats>
+</seelog>`
+	return c
+}
+
 // SetLevel sets the log level for logging
 func SetLevel(logLevel string) {
+	levels := map[string]string{
+		"debug": "debug",
+		"info":  "info",
+		"warn":  "warn",
+		"error": "error",
+		"crit":  "critical",
+		"none":  "off",
+	}
 	parsedLevel, ok := levels[strings.ToLower(logLevel)]
 
 	if ok {
-		levelLock.Lock()
-		defer levelLock.Unlock()
-		level = parsedLevel
+		Config.lock.Lock()
+		defer Config.lock.Unlock()
+		Config.level = parsedLevel
 		reloadConfig()
 	}
 }
 
 // GetLevel gets the log level
 func GetLevel() string {
-	levelLock.RLock()
-	defer levelLock.RUnlock()
+	Config.lock.Lock()
```

Comment:
Isn't read lock good enough here since we are just reading here?

Reply:
yes but I am using just a plain sync.Mutex....a RWMutex seemed like overkill given the low-throughput expectation here